### PR TITLE
[PUBDEV-6922] Send data to and from external backend not by each item, but after specified block size

### PR DESCRIPTION
**Pull request by jakubhava**
The goal of this change is to send the data to and from external backend after some pre-defined ( currently hard-coded) batch size to improve the overall performance

### CODE REVIEW

**Comment by mn-mikke** [comment]
This feature improves performance of conversion Spark DataFrames to H2O Frames with external backend significantly:
```
Tested on EMR 4+4 nodes (1 core each)

before:
300M rows, 1 column:    312225 milliseconds
300M rows, 10 columns: 1527652 milliseconds
300M rows, 100 columns: timeout after 4 hours

after:
300M rows, 1 column:     46142 milliseconds
300M rows, 10 columns:  111857 milliseconds
300M rows, 100 columns: 630779 milliseconds
```

File: h2o-core/src/main/java/water/ExternalFrameWriterClient.java [hunk 1]
```diff
@@ -60,6 +59,7 @@
   private int timeout;
   private int numRows;
   private int numCols;
+  private long blockSize;
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ExternalFrameWriterClient.java]: Minor: these could be final ?

File: h2o-core/src/main/java/water/ExternalFrameWriterClient.java [hunk 2]
```diff
@@ -124,77 +125,79 @@ public void finalizeFrame(String keyName, long[] rowsPerChunk, byte[] colTypes,
...
 
-    public void close() throws IOException, ExternalFrameConfirmationException {
-        waitForRequestToFinish(timeout, ExternalBackendRequestType.WRITE_TO_CHUNK.getByte());
-        channel.close();
-    }
+  public void close() throws IOException, ExternalFrameConfirmationException {
+    // write remaining data
+    writeToChannel(ab, channel);
+    waitForRequestToFinish(timeout, ExternalBackendRequestType.WRITE_TO_CHUNK.getByte());
+    channel.close();
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/ExternalFrameWriterClient.java]: What if the wait operation times out ? Is the channel.close() is going to be closed and resources freed ?

File: h2o-core/src/main/java/water/ExternalFrameWriterClient.java (hunk 1 above)
Comment (mn-mikke) [reply h2o-core/src/main/java/water/ExternalFrameWriterClient.java]: Good point, thanks!

**Comment by mn-mikke** [comment]
@Pscheidl I've tried to address your changes. Please take a look, thanks!

**Comment by Pscheidl** [comment]
I think it's great. Thanks.

**Comment by jakubhava** [comment]
Great speedup, thanks for measuring Marek. Did you did manual benchmarks or do we have these already in our benchmarks suite on aws?

**Pull request merged**